tss2-dlopen/tss2-dlopen-esys.c

```diff
@@ -0,0 +1,2171 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*******************************************************************************
+ * Copyright 2020, Fraunhofer SIT
+ * All rights reserved. *******************************************************************************/
+
+/**
+ * The purpose of this file is to copy it into your project and
+ * include it during compilation if you don't want to link against
+ * libtss2-esys at compile time.
+ * It will attempt to load libtss2-esys.so during runtime.
+ * It will either work similarly to directly linking to libtss2-esys.so
+ * at compile-time or return a NOT_IMPLEMENTED error.
+ *
+ * For new versions of this file, please check:
+ * http://github.com/tpm2-software/tpm2-tss/tss2-dlopen
+*/
+
+#include <dlfcn.h>
+#include <stdio.h>
+#include <tss2/tss2_esys.h>
+
+#define str(s) xstr(s)
+#define xstr(s) #s
+
+static void *esys_dlhandle = NULL;
+
+TSS2_RC
+Esys_Initialize(
+    ESYS_CONTEXT **esys_context,
+    TSS2_TCTI_CONTEXT *tcti,
+    TSS2_ABI_VERSION *abiVersion)
+{
+    if (!esys_dlhandle) {
+        esys_dlhandle = dlopen("tss2_esys", RTLD_NOW & RTLD_LOCAL);
+        if (!esys_dlhandle) {
+            fprintf(stderr, "WARNING: Library libtss2_esys not found.");
+            return TSS2_ESYS_RC_NOT_IMPLEMENTED;
+        }
+    }
+
+    TSS2_RC (*sym) (ESYS_CONTEXT **esys_context, TSS2_TCTI_CONTEXT *tcti, TSS2_ABI_VERSION *abiVersion) = dlsym(esys_dlhandle, "Esys_Initialize");
+    if (!sym) {
+        fprintf(stderr, "WARNING: Function Esys_Initialize not found.");
+        return TSS2_ESYS_RC_NOT_IMPLEMENTED;
+    }
+
+    return sym(esys_context, tcti, abiVersion);
+}
+
+void
+Esys_Finalize(ESYS_CONTEXT **ctx)
+{
+    TSS2_RC (*sym) (ESYS_CONTEXT **ctx) = dlsym(esys_dlhandle, "Esys_Finalize");
+    if (!sym) {
+        fprintf(stderr, "WARNING: Function Esys_Finalize not found.");
+        return;
+    }
+    sym(ctx);
+}
+
+void
+Esys_Free(void *__ptr)
+{
+    TSS2_RC (*sym) (void *__ptr) = dlsym(esys_dlhandle, "Esys_Free");
+    if (!sym) {
+        fprintf(stderr, "WARNING: Function Esys_Free not found.");
+        return;
+    }
+    sym(__ptr);
+}
+
+#define MAKE_ESYS_0(fun) \
+TSS2_RC fun (ESYS_CONTEXT *ctx) { \
+    TSS2_RC (*sym) (ESYS_CONTEXT *ctx) = \
+        dlsym(esys_dlhandle, str(fun)); \
+    if (!sym) { \
+        fprintf(stderr, "WARNING: Function " str(fun) " not found."); \
```

Comment:
Does it really matter ?
I just took the same string was for symbol lookup.

Could be `__func__` there as well, I agree but maybe someone wants to rename the function names to eg `dl_ ## fun` and then it would not be correct here anymore.